ffi/src/expressions.rs

```diff
@@ -245,3 +249,417 @@ pub extern "C" fn visit_expression_literal_bool(
+            let value_id = visit_scalar(visitor, value);
+            call!(
+                visitor,
+                visit_struct_literal_field,
+                struct_id,
+                field.name().into(),
+                value_id
+            );
+        }
+        struct_id
+    }
+    fn visit_struct(visitor: &mut EngineExpressionVisitor, exprs: &Vec<Expression>) -> usize {
+        let expr_struct_id = call!(visitor, visit_struct, exprs.len());
+        for expr in exprs {
+            let expr_id = visit_expression(visitor, expr);
+            call!(visitor, visit_struct_sub_expr, expr_struct_id, expr_id)
+        }
+        expr_struct_id
+    }
+    fn visit_variadic(
+        visitor: &mut EngineExpressionVisitor,
+        op: &VariadicOperator,
+        exprs: &Vec<Expression>,
+    ) -> usize {
+        let visit_fn = match op {
+            VariadicOperator::And => &visitor.visit_and,
+            VariadicOperator::Or => &visitor.visit_or,
+        };
+        let variadic_id = visit_fn(visitor.data, exprs.len());
+        for expr in exprs {
+            let expr_id = visit_expression(visitor, expr);
+            call!(visitor, visit_variadic_sub_expr, variadic_id, expr_id)
+        }
+        variadic_id
+    }
+    fn visit_scalar(visitor: &mut EngineExpressionVisitor, scalar: &Scalar) -> usize {
+        match scalar {
+            Scalar::Integer(val) => call!(visitor, visit_int, *val),
+            Scalar::Long(val) => call!(visitor, visit_long, *val),
+            Scalar::Short(val) => call!(visitor, visit_short, *val),
+            Scalar::Byte(val) => call!(visitor, visit_byte, *val),
+            Scalar::Float(val) => call!(visitor, visit_float, *val),
+            Scalar::Double(val) => call!(visitor, visit_double, *val),
+            Scalar::String(val) => call!(visitor, visit_string, val.into()),
+            Scalar::Boolean(val) => call!(visitor, visit_bool, *val),
+            Scalar::Timestamp(val) => call!(visitor, visit_timestamp, *val),
+            Scalar::TimestampNtz(val) => call!(visitor, visit_timestamp_ntz, *val),
+            Scalar::Date(val) => call!(visitor, visit_date, *val),
+            Scalar::Binary(buf) => call!(visitor, visit_binary, buf.as_ptr(), buf.len()),
+            Scalar::Decimal(value, precision, scale) => {
+                let ms: u64 = (value >> 64) as u64;
+                let ls: u64 = *value as u64;
+                call!(visitor, visit_decimal, ms, ls, *precision, *scale)
+            }
+            Scalar::Null(_) => call!(visitor, visit_null),
+            Scalar::Struct(struct_data) => visit_struct_literal(visitor, struct_data),
+            Scalar::Array(array) => visit_array(visitor, array),
+        }
+    }
+    fn visit_expression(visitor: &mut EngineExpressionVisitor, expression: &Expression) -> usize {
+        match expression {
+            Expression::Literal(scalar) => visit_scalar(visitor, scalar),
```

Comment:
Cool, I can add that.

Reply:
Ah sry, I'd focused on the `visit_literal(scalar_id: usize)` approach. Giving it more thought, I'm not sure if adding the extra layer helps much. `visit_int` will always be a scalar, so perhaps it's best to avoid the extra FFI calls of a separate `visit_int` + `visit_literal`, and just call it `visit_int_literal`.  Are there any specific use cases you see for the separate `visit_literal`?

Reply:
currently no. I think `visit_int_literal` is probably fine for now yeah

Reply:
Small suggestion tho -- `visit_literal_xxx` method names might be easier to grok and group visually?

Reply:
Done 👍